docs/website/docs/v1.x/dsl/middleware.md

```diff
@@ -1,3 +1,217 @@
 # Middleware
 
-WIP
\ No newline at end of file
+Middlewares are transformations that one can apply on any Http to produce a new one. 
+They can modify requests and responses and also transform them into more concrete domain entities.
+
+Middleware is simply a function that takes one `Http` as a parameter and returns another `Http` `f(Http) => Http` elaborated below in terms of type parameters "In" and "Out"
+
+```scala
+type Middleware[R, E, AIn, BIn, AOut, BOut] = Http[R, E, AIn, BIn] => Http[R, E, AOut, BOut]
+```
+
+* `AIn` and `BIn` are type params of the input `Http`
+* `AOut` and `BOut` are type params of the output `Http`
+
+## Create Middleware
+
+### Middleware that does nothing
+
+returns the same `Http` as input without doing any modification
+
+```scala
+val middleware: Middleware[Any, Nothing, Nothing, Any, Any, Nothing] = Middleware.identity
+```
+
+### Middleware that always succeeds
+
+creates a middleware that always returns the output `Http` that succeeds with the given value and never fails.
+
+```scala
+val middleware: Middleware[Any, Nothing, Nothing, Any, Any, Int] = Middleware.succeed(1)
+```
+
+### Middleware that always fails
+
+creates a middleware that always returns the output `Http` that always fails.
+
+```scala
+val middleware: Middleware[Any, String, Nothing, Any, Any, Nothing] = Middleware.fail("error")
+```
+
+### Middleware from a partial function
+
+- `collect` creates middleware using a specified function
+
+```scala
+val middleware: Middleware[Any, Nothing, Request, Response, Request, Response] = Middleware.collect[Request](_ => Middleware.addHeaders(Headers("a", "b")))
+```
+
+- `collectZIO` creates middleware using a specified effect function
+
+```scala
+val middleware: Middleware[Any, Nothing, Request, Response, Request, Response] = Middleware.collectZIO[Request](_ => ZIO.succeed(Middleware.addHeaders(Headers("a", "b"))))
+```
+
+### Middleware using codec
+
+codec takes two functions `decoder: AOut => Either[E, AIn]` and `encoder: BIn => Either[E, BOut]`
+
+The below snippet takes two functions:
+- decoder function to decode Request to String 
+- encoder function to encode String to Response
+
```

Comment:
and will create a middleware that gives out an `Http[R,E,Request,Response]` from `Http[R,E,String,String]`